presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,283 @@
+        {
+            checkArgument(totalNodes > 0, "totalNodes must be greater than 0");
+            Set<Integer> numbers = IntStream.range(0, totalNodes)
+                    .boxed()
+                    .collect(toImmutableSet());
+            return Sets.powerSet(numbers).stream()
+                    .filter(subSet -> subSet.contains(0))
+                    .filter(subSet -> subSet.size() < numbers.size());
+        }
+
+        public static JoinNode createJoinAccordingToPartitioning(JoinGraphNode joinGraph, Set<Integer> partitioning, PlanNodeIdAllocator idAllocator)
+        {
+            List<PlanNode> sources = joinGraph.getSources();
+            List<PlanNode> leftSources = partitioning.stream()
+                    .map(sources::get)
+                    .collect(toImmutableList());
+            List<PlanNode> rightSources = ImmutableList.copyOf(Sets.difference(ImmutableSet.copyOf(sources), ImmutableSet.copyOf(leftSources)));
+            Set<Symbol> leftSymbols = leftSources.stream()
+                    .flatMap(node -> node.getOutputSymbols().stream())
+                    .collect(toImmutableSet());
+
+            InnerJoinPredicateUtils.InnerJoinPushDownResult pushDownResult = sortPredicatesForInnerJoin(leftSymbols, ImmutableList.of(joinGraph.getFilter()), ImmutableList.of());
+
+            Set<Symbol> joinSymbols = ImmutableSet.<Symbol>builder()
+                    .addAll(joinGraph.getOutputSymbols())
+                    .addAll(DependencyExtractor.extractUnique(pushDownResult.getJoinPredicate()))
+                    .build();
+            PlanNode left = getJoinSource(
+                    idAllocator,
+                    leftSources,
+                    pushDownResult.getLeftPredicate(),
+                    joinSymbols.stream().filter(leftSymbols::contains).collect(toImmutableList()));
+            PlanNode right = getJoinSource(
+                    idAllocator,
+                    rightSources,
+                    pushDownResult.getRightPredicate(),
+                    joinSymbols.stream()
+                            .filter(symbol -> !leftSymbols.contains(symbol))
+                            .collect(toImmutableList()));
+
+            List<Expression> joinPredicates = extractConjuncts(pushDownResult.getJoinPredicate());
+            List<JoinNode.EquiJoinClause> joinConditions = joinPredicates.stream()
+                    .filter(JoinEnumerator::isJoinEqualityCondition)
+                    .map(predicate -> toEquiJoinClause((ComparisonExpression) predicate, leftSymbols))
+                    .collect(toImmutableList());
+            List<Expression> joinFilters = joinPredicates.stream()
+                    .filter(predicate -> !isJoinEqualityCondition(predicate))
+                    .collect(toImmutableList());
+
+            // sort output symbols so that the left input symbols are first
+            List<Symbol> inputSymbols = ImmutableList.<Symbol>builder()
```

Comment:
you don't need to materialize `inputSymbols`. Make it prefix of pipeline that produces `outputSymbols` e.g:
```
Stream.concat(left.getOutputSymbols(), right.getOutputSymbols())
  .filter(joinGraph.getOutputSymbols()::contains)
  .collect(toImmutableList());
```